mitodl/mit-open
# Improve contentfile api query performance

### What are the relevant tickets?
Closes #386

### Description (What does it do?)
- Adds a database index for the `created_on` field of the abstract `TiimeStampedModel` (leading to migrations for all models inheriting from it).
- Modifies the `indexing_api.deindex_run_content_files` function to delete unpublished contentfiles for the run after they have been deindexed, since there is no longer any use for them.  If the run ever gets published again, they will be reimported.
- Changes the default sort field for `ContentFileViewSet` from `-updated_on` to `-created_on` (more comparable to the `new` viewset for learning resources).  Though it would probably be fine to just remove this sort altogether and add a `new` viewset if needed in the future.  I can make that change if desired.
- Fixes a bug in calling the `SearchIndexPlugin` for learning resource runs that I noticed during testing, and adds a bunch of tests for that function and other related ones.

I thought about adding an index to `updated_on` as well, ultimately decided against it for now since it is not currently used for sorting (after the change in this PR).  But willing to change my mind on that.


### How can this be tested?
You can run the following queries before and after applying the migrations.  After the migrations, sorting by `-created_on` should be significantly faster than sorting by `-updated_on`

```python
import time
from learning_resources.models import *

start = time.time();print(ContentFile.objects.filter(published=True).order_by("-created_on")[:20].values_list("id", flat=True));end = time.time();print(end - start)

<TimestampedModelQuerySet [103249, 103248, 103247, 103246, 103245, 103244, 103243, 103242, 103241, 103240, 103239, 103238, 103237, 103236, 103235, 103234, 103233, 103232, 103231, 103230]>
0.03517913818359375

start = time.time();print(ContentFile.objects.filter(published=True).order_by("-updated_on")[:20].values_list("id", flat=True));end = time.time();print(end - start)

<TimestampedModelQuerySet [103249, 24855, 103248, 24853, 24852, 24851, 24850, 24849, 24848, 24847, 24846, 24845, 24844, 24843, 24842, 24841, 103247, 24839, 103246, 103245]>
0.5206596851348877

start = time.time();print(ContentFile.objects.filter(published=True)[:20].values_list("id", flat=True));end = time.time();print(end - start)

<TimestampedModelQuerySet [77994, 77995, 77996, 77997, 77998, 89129, 1856, 1859, 89126, 89127, 89128, 89130, 89131, 89132, 89133, 90226, 90227, 90228, 90229, 90230]>
0.0017976760864257812
```
